### PR TITLE
[b/354721825] Add Apache Ranger connector.

### PR DESCRIPTION
**Pull request by aleofreddi**

### CODE REVIEW

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerClient.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+  public ImmutableList<Policy> findPolicies(Map<String, String> args) throws RangerException {
+    return doHttpGet("/service/plugins/policies", args, ListPoliciesResponse.class).policies();
+  }
+
+  private <T> T doHttpGet(String path, Map<String, String> params, Class<T> bodyClass)
+      throws RangerException {
+    URI uri;
+    try {
+      uri =
+          new URIBuilder(baseUri + path)
```
Comment (dawidxc) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerClient.java:173]: I would suggest to either do:
```
new URIBuilder(baseUri).setPath(path)
```

or if the baseUri can also contain path, then add a static method that would concatenate `baseUri` and `path` making sure that a single slash is added in between no matter if they start/end with a slash or not.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerConnector.java [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+    return ArchiveNameUtil.getFileName(getName());
+  }
+
+  @Override
+  public void addTasksTo(List<? super Task<?>> out, @Nonnull ConnectorArguments arguments) {
+    out.add(new DumpGroupsTask());
+    out.add(new DumpPoliciesTask());
+    out.add(new DumpRolesTask());
+    out.add(new DumpServicesTask());
+    out.add(new DumpUsersTask());
```
Comment (dawidxc) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerConnector.java:90]: I would suggest adding standard dumper metadata tasks - `DumpMetadataTask` and `FormatTask`. Possibly in a separate PR.

https://github.com/google/dwh-migration-tools/blob/42e48dc2c8d6e2e1220373ace09018059f88a9c8/dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hadoop/HadoopMetadataConnector.java#L67-L68

File: dumper/app/src/test/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerPageIteratorTest.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+
+@RunWith(JUnit4.class)
+public class RangerPageIteratorTest extends TestCase {
```
Comment (dawidxc) [inline dumper/app/src/test/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerPageIteratorTest.java:40]: I would suggest adding a test for an empty list.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerClient.java [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+      throw new RangerException("Failed to build API URI", e);
+    }
+    HttpGet httpRequest = new HttpGet(uri);
+    httpRequest.addHeader(ACCEPT, APPLICATION_JSON.toString());
+    try {
+      httpRequest.addHeader(new BasicScheme().authenticate(credentials, httpRequest));
+    } catch (AuthenticationException e) {
+      throw new RangerException("Failed to initialize authentication header", e);
+    }
+    CloseableHttpResponse httpResponse = null;
```
Comment (paolomorandini) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerClient.java:189]: Should this be moved in a `t-w-r` to ensure it will get closed?

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerConnector.java [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+    public AbstractRangerTask(String targetPath) {
+      super(targetPath);
+    }
+
+    @ForOverride
+    protected abstract Iterator<T> dataIterator(
+        @Nonnull Writer writer, @Nonnull RangerClientHandle handle) throws Exception;
+
+    @Override
+    protected Void doRun(TaskRunContext context, ByteSink sink, @Nonnull Handle handle)
```
Comment (paolomorandini) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerConnector.java:244]: Consider adding nullability annotations to all arguments or only to `@Nullable` ones.

File: dumper/lib-dumper-spi/src/main/java/com/google/edwmigration/dumper/plugin/lib/dumper/spi/RangerDumpFormat.java [hunk 6]
```diff
@@ -0,0 +1,1073 @@
...
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
+import com.google.auto.value.AutoValue;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.time.Instant;
+import javax.annotation.Nullable;
+
+public interface RangerDumpFormat {
+
+  ObjectMapper MAPPER =
```
Comment (paolomorandini) [inline dumper/lib-dumper-spi/src/main/java/com/google/edwmigration/dumper/plugin/lib/dumper/spi/RangerDumpFormat.java:36]: Can we have only one `MAPPER` (the other one is in `RangerClient`)?

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerConnector.java (hunk 2 above)
Comment (aleofreddi) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerConnector.java:90]: Good point, I'll add these in a subsequent PR.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerClient.java (hunk 4 above)
Comment (aleofreddi) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerClient.java]: Good catch, this would have definitely leaked connections.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerClient.java (hunk 1 above)
Comment (aleofreddi) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerClient.java]: Done.

File: dumper/lib-dumper-spi/src/main/java/com/google/edwmigration/dumper/plugin/lib/dumper/spi/RangerDumpFormat.java (hunk 6 above)
Comment (aleofreddi) [reply dumper/lib-dumper-spi/src/main/java/com/google/edwmigration/dumper/plugin/lib/dumper/spi/RangerDumpFormat.java:36]: I've kept them distinct because the one in `RangerClient` is solely used to (de)serialize Ranger API payloads, whereas the one in `RangerDumpFormat` is specifically used for serialization into the zip file. Each has its own set of settings.

Probably they could be merged in one, but I think it makes sense to keep these separate as some settings are only meaningful in one context (for example, `FAIL_ON_UNKNOWN_PROPERTIES` makes sense only when reading from Ranger API).

**Review by paolomorandini — APPROVED** [review]

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerClient.java [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+import org.apache.http.message.BasicNameValuePair;
+import org.apache.http.util.EntityUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RangerClient {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RangerConnector.class);
+
+  private static final ObjectMapper MAPPER =
```
Comment (misolt) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerClient.java:59]: It looks like this `ObjectMapper` is mutable, but the uppercase name `MAPPER` suggests a constant. I'd recommend renaming it to `objectMapper` or `mapper`.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerConnector.java [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+      super(targetPath);
+    }
+
+    @ForOverride
+    protected abstract Iterator<T> dataIterator(
+        @Nonnull Writer writer, @Nonnull RangerClientHandle handle) throws Exception;
+
+    @Override
+    protected Void doRun(
+        @Nonnull TaskRunContext context, @Nonnull ByteSink sink, @Nonnull Handle handle)
```
Comment (misolt) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerConnector.java:246]: Here I'd remove the `@Nonnull` before `TaskRunContext`, because the superclass method accepts a nullable `context` parameter:

https://github.com/google/dwh-migration-tools/blob/934ed139dc25284144f9ed384c0c91fadcc8679c/dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/task/AbstractTask.java#L105

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerClient.java (hunk 7 above)
Comment (aleofreddi) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerClient.java:59]: Done.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerConnector.java (hunk 8 above)
Comment (aleofreddi) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/ranger/RangerConnector.java:246]: Agree, done.
